app/src/main/java/ch/epfl/qedit/backend/database/FirebaseDBService.java

```diff
@@ -0,0 +1,54 @@
+package ch.epfl.qedit.backend.database;
+
+import androidx.annotation.NonNull;
+import ch.epfl.qedit.model.Question;
+import ch.epfl.qedit.model.Quiz;
+import ch.epfl.qedit.util.BundledData;
+import ch.epfl.qedit.util.Callback;
+import ch.epfl.qedit.util.Response;
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.firestore.DocumentSnapshot;
+import com.google.firebase.firestore.FirebaseFirestore;
+import java.util.ArrayList;
+
+public class FirebaseDBService implements DatabaseService {
+
+    private FirebaseFirestore db;
+
+    public FirebaseDBService() {
+        // Access a Cloud Firestore instance
+        db = FirebaseFirestore.getInstance();
+    }
+
+    // TODO this function will extract Quiz from a document coming from firestore
+    private BundledData getQuizFromDocument(DocumentSnapshot document) {
+        return new Quiz(new ArrayList<Question>()).toBundle();
+    }
+
+    @Override
+    public void getBundle(
+            final String collection,
+            String document,
+            final Callback<Response<BundledData>> responseCallback) {
+        db.collection(collection)
+                .document(document)
+                .get()
+                .addOnCompleteListener(
+                        new OnCompleteListener<DocumentSnapshot>() {
+                            @Override
+                            public void onComplete(@NonNull Task<DocumentSnapshot> task) {
+                                Response<BundledData> response;
+                                if (task.isSuccessful()) {
+                                    DocumentSnapshot document = task.getResult();
+                                    if (document.exists())
+                                        if (collection.equals("quizzes"))
+                                            response = Response.ok(getQuizFromDocument(document));
```

Comment:
Just as a helper: I created the `BundledData` class with firestore in mind. This means that the data from the document can directly be extracted by doing `new BundledData(document.getData())`.
It is then the responsibility of the caller of the database query to transform the data into a Java type. Therefore you don't need to interpret the data in the database implementation, but rather just return it.

Reply:
To clarify, it means that you have to write:
`response = Response.ok(new BundledData(document.getData());`
instead of what you wrote.